Framework/Algorithms/src/ReflectometryReductionOneAuto3.cpp

```diff
@@ -0,0 +1,1126 @@
+    const std::string IvsLamName = outputIvsLamNames[i];
+
+    // Find the spectrum processing instructions for ws index 0
+    auto currentWorkspace = boost::dynamic_pointer_cast<MatrixWorkspace>(
+        AnalysisDataService::Instance().retrieve(outputIvsLamNames[i]));
+    auto newProcInst = convertToSpectrumNumber("0", currentWorkspace);
+    alg->setProperty("ProcessingInstructions", newProcInst);
+    alg->setProperty("InputWorkspace", IvsLamName);
+    alg->setProperty("OutputWorkspace", IvsQName);
+    alg->setProperty("OutputWorkspaceBinned", IvsQBinnedName);
+    alg->setProperty("OutputWorkspaceWavelength", IvsLamName);
+    alg->execute();
+    IvsQBinnedGroup.push_back(IvsQBinnedName);
+    IvsQGroup.push_back(IvsQName);
+    if (AnalysisDataService::Instance().doesExist(IvsLamName)) {
+      IvsLamGroup.push_back(IvsLamName);
+    }
+  }
+
+  setOutputWorkspaces(output, IvsLamGroup, IvsQBinnedGroup, IvsQGroup);
+
+  return true;
+}
+
+/** Construct a polarization efficiencies workspace based on values of input
+ * properties.
+ */
+std::tuple<API::MatrixWorkspace_sptr, std::string, std::string>
+ReflectometryReductionOneAuto3::getPolarizationEfficiencies() {
+  auto groupIvsLam = AnalysisDataService::Instance().retrieveWS<WorkspaceGroup>(
+      getPropertyValue("OutputWorkspaceWavelength"));
+
+  Workspace_sptr workspace = groupIvsLam->getItem(0);
+
+  auto effAlg = createChildAlgorithm("ExtractPolarizationEfficiencies");
+  effAlg->setProperty("InputWorkspace", workspace);
+  effAlg->execute();
+  MatrixWorkspace_sptr efficiencies = effAlg->getProperty("OutputWorkspace");
+  std::string correctionMethod = effAlg->getPropertyValue("CorrectionMethod");
+  std::string correctionOption = effAlg->getPropertyValue("CorrectionOption");
+
+  return std::make_tuple(efficiencies, correctionMethod, correctionOption);
+}
+
+/**
+ * Apply a polarization correction to workspaces in lambda.
+ * @param outputIvsLam :: Name of a workspace group to apply the correction
+ * to.
+ */
+void ReflectometryReductionOneAuto3::applyPolarizationCorrection(
+    std::string const &outputIvsLam) {
+  MatrixWorkspace_sptr efficiencies;
+  std::string correctionMethod;
+  std::string correctionOption;
+  std::tie(efficiencies, correctionMethod, correctionOption) =
+      getPolarizationEfficiencies();
+  CorrectionMethod::validate(correctionMethod);
+
+  Algorithm_sptr polAlg = createChildAlgorithm("PolarizationEfficiencyCor");
+  polAlg->setChild(false);
+  polAlg->setRethrows(true);
+  polAlg->setProperty("OutputWorkspace", outputIvsLam);
+  polAlg->setProperty("Efficiencies", efficiencies);
+  polAlg->setProperty("CorrectionMethod", correctionMethod);
+  polAlg->setProperty(CorrectionMethod::OPTION_NAME.at(correctionMethod),
+                      correctionOption);
+
+  if (correctionMethod == "Fredrikze") {
+    polAlg->setProperty("InputWorkspaceGroup", outputIvsLam);
+    polAlg->execute();
+  } else {
```

Comment:
I think it might be better if we else if Wildes and then raise some error in the `else`. This could help avoid errors caused by forgetting to include a new correctionMethod in every instance it's used in the algorithm?

Reply:
An error would have been raised by `CorrectionMethod::validate(correctionMethod);` on line 1006 if the correction method is anything other than Wildes or Fredrikze so I'm not sure it's necessary.